test/common/upstream/health_check_fuzz.cc

```diff
@@ -0,0 +1,340 @@
+  cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+      makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  expectSessionCreate();
+  expectClientCreate();
+  health_checker_->start();
+  if (input.health_check_config().has_reuse_connection()) {
+    reuse_connection_ = input.health_check_config().reuse_connection().value();
+  }
+
+  if (DurationUtil::durationToMilliseconds(input.health_check_config().initial_jitter()) != 0) {
+    interval_timer_->invokeCallback();
+  }
+}
+
+void TcpHealthCheckFuzz::respond(std::string data, bool last_action) {
+  if (!timeout_timer_->enabled_) {
+    ENVOY_LOG_MISC(trace, "Timeout timer is disabled. Skipping response.");
+    return;
+  }
+  Buffer::OwnedImpl response;
+  response.add(data);
+
+  ENVOY_LOG_MISC(trace, "Responded with {}. Length (in bytes) = {}. This is the string passed in.",
+                 data, data.length());
+  read_filter_->onData(response, true);
+
+  // The interval timer may not be on. If it's not on, return. An http response will automatically
+  // turn on interval and turn off timer, but for tcp it doesn't if the data doesn't match. If it
+  // doesn't match, it only sets the host to unhealthy. If it doesn't hit, it will turn timeout on
+  // and interval off.
+  if (!reuse_connection_ && !last_action && interval_timer_->enabled_) {
+    expectClientCreate();
+    interval_timer_->invokeCallback();
+  }
+}
+
+void TcpHealthCheckFuzz::triggerIntervalTimer() {
+  if (!interval_timer_->enabled_) {
+    ENVOY_LOG_MISC(trace, "Interval timer is disabled. Skipping trigger interval timer.");
+    return;
+  }
+  ENVOY_LOG_MISC(trace, "Triggered interval timer");
+  interval_timer_->invokeCallback();
+}
+
+void TcpHealthCheckFuzz::triggerTimeoutTimer(bool last_action) {
+  if (!timeout_timer_->enabled_) {
+    ENVOY_LOG_MISC(trace, "Timeout timer is disabled. Skipping trigger timeout timer.");
+    return;
+  }
+  ENVOY_LOG_MISC(trace, "Triggered timeout timer");
+  timeout_timer_->invokeCallback(); // This closes the client, turns off timeout
+                                    // and enables interval
+  if (!last_action) {
+    ENVOY_LOG_MISC(trace, "Creating client and stream from network timeout");
+    expectClientCreate();
+    interval_timer_->invokeCallback();
+  }
+}
+
+void TcpHealthCheckFuzz::raiseEvent(const Network::ConnectionEvent& event_type, bool last_action) {
+  connection_->raiseEvent(event_type);
+  if (!last_action && event_type != Network::ConnectionEvent::Connected) {
+    if (!interval_timer_
+             ->enabled_) { // Note: related to the TODO below. This will mean that both timers are
+                           // disabled, meaning any action after hitting this will be voided.
+      return;
+    }
+    ENVOY_LOG_MISC(trace, "Creating client from close event");
+    expectClientCreate();
+    // It calls it with a number if expect close is false (like 4 different code paths). TODO:
```

Comment:
Calls it with a number means it calls interval timer with a number. This comment is left over from when I thought calling interval timer with enable timer wouldnt call enabled_. However (_, _) in gmock meant that it can be called with anything. Investigating this logic (and if can actually create a state where both timers are disabled)

Reply:
I figured out whats happening here. First, since I have structured my fuzzer to always create a client by invoking the interval timer after any event that would close it, such as a timeout, or a response with reuse connection set to false in the config. Thus, the codepaths that would set expect close to be false and thus always hit the conditional that calls onFailure(). This is only time where the boolean would be questionable is this line specifically: https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/health_checker_impl.cc#L489. Looking into it now (...Feedback... :P)

Reply:
Okay so I induced a seg fault in that specific case, fixing now.

Reply:
Okay I fixed it, you'll see what I did in the next commit. Let me know what you think!

Reply:
Basically, I checked if it hit that path in my fuzzer (not using any state from health checker), and created a client then, similar to how I would create client in other scenarios ^^^ that would set expect_close_ to false.